recipes/node-url-to-whatwg-url/src/url-parse.ts

```diff
@@ -0,0 +1,226 @@
+import type { SgRoot, Edit } from "@codemod.com/jssg-types/main";
+import { getNodeImportStatements } from "@nodejs/codemod-utils/ast-grep/import-statement";
+import { getNodeRequireCalls } from "@nodejs/codemod-utils/ast-grep/require-call";
+import { resolveBindingPath } from "@nodejs/codemod-utils/ast-grep/resolve-binding-path";
+
+/**
+ * Transforms `url.parse` usage to `new URL()`.
+ *
+ * See https://nodejs.org/api/deprecations.html#DEP0116 for more details.
+ *
+ * Handle:
+ * 1. `url.parse(urlString)` → `new URL(urlString)`
+ * 2. `parse(urlString)` → `new URL(urlString)`
+ * if imported with aliases
+ * 2. `foo.parse(urlString)` → `new URL(urlString)`
+ * 3. `foo(urlString)` → `new URL(urlString)`
+ */
+export default function transform(root: SgRoot): string | null {
+    const rootNode = root.root();
+    const edits: Edit[] = [];
+
+    // Safety: only run on files that import/require node:url
+    const hasNodeUrlImport =
+        getNodeImportStatements(root, "url").length > 0 ||
+		getNodeRequireCalls(root, "url").length > 0;
+
+	if (!hasNodeUrlImport) return null;
+
+	// 1) Replace parse calls with new URL() using binding-aware patterns
+    const importNodes = getNodeImportStatements(root, "url");
+    const requireNodes = getNodeRequireCalls(root, "url");
+	const parseCallPatterns = new Set<string>();
+
+	for (const node of [...importNodes, ...requireNodes]) {
+        const binding = resolveBindingPath(node, "$.parse");
+        if (!binding) continue;
+        parseCallPatterns.add(`${binding}($ARG)`);
+    }
+
+	// 1.a) Identify variables assigned from parse(...) so we only rewrite legacy
+	// properties (auth, path, hostname) on those specific objects
+	const parseResultVars = new Set<string>();
+	for (const pattern of parseCallPatterns) {
+		const matches = rootNode.findAll({
+			rule: {
+				any: [
+					{ pattern: `const $OBJ = ${pattern}` },
+					{ pattern: `let $OBJ = ${pattern}` },
+					{ pattern: `var $OBJ = ${pattern}` },
+					{ pattern: `$OBJ = ${pattern}` }
+				]
+			}
+		});
+		for (const m of matches) {
+			const obj = m.getMatch("OBJ");
+			if (!obj) continue;
+			const name = obj.text();
+			if (/^[$A-Z_a-z][$\w]*$/.test(name)) parseResultVars.add(name);
+		}
+	}
+
+	// 1.b) Replace parse calls with new URL()
+	//      Also, for declarations using `var`, upgrade to `let` while keeping `const` as-is.
```

Comment:
I think that upgrade from `var` → `let` is not safe: `var`'s scoping is different than `let`

```js
function foo(a) {
  switch(a) {
    case '…': {
      var u = url.parse(…);
      let l = u;
    }
  }

  u.slice(0,10); // works
  l.slice(0,10); // 💥
}
```